benefits/enrollment/views.py

```diff
@@ -84,25 +87,81 @@ def index(request):
         client.oauth.ensure_active_token(client.token)
 
         funding_source = client.get_funding_source_by_token(card_token)
+        funding_source_id = funding_source.id
+        group_id = eligibility.group_id
 
         try:
-            client.link_concession_group_funding_source(funding_source_id=funding_source.id, group_id=eligibility.group_id)
+            group_funding_source = _get_group_funding_source(
+                client=client, group_id=group_id, funding_source_id=funding_source_id
+            )
+
+            already_enrolled = group_funding_source is not None
+            has_no_expiration_date = already_enrolled and group_funding_source.concession_expiry is None
+            has_expiration_date = already_enrolled and group_funding_source.concession_expiry is not None
+
+            if eligibility.supports_expiration:
+                # set expiry on session
+                if has_expiration_date:
+                    session.update(request, enrollment_expiry=group_funding_source.concession_expiry)
+                else:
+                    session.update(request, enrollment_expiry=_calculate_expiry(eligibility.expiration_days))
+
+                if not already_enrolled:
+                    # enroll user with an expiration date, return success
+                    _enroll(
+                        client=client,
+                        group_id=group_id,
+                        funding_source_id=funding_source_id,
+                        expiry_date=session.enrollment_expiry(request),
+                    )
+                    return _success(request, group_id)
+                else:
+                    if has_no_expiration_date:
+                        # update expiration of existing enrollment, return success
+                        _enroll(
+                            client,
+                            group_id=group_id,
+                            funding_source_id=funding_source_id,
+                            expiry_date=session.enrollment_expiry(request),
+                        )
+                        return _success(request, group_id)
+                    else:
+                        is_expired = _is_expired(group_funding_source.concession_expiry)
+                        is_within_reenrollment_window = _is_within_reenrollment_window(
+                            group_funding_source.concession_expiry, session.enrollment_reenrollment(request)
+                        )
+
+                        if is_expired or is_within_reenrollment_window:
+                            # update expiration of existing enrollment, return success
+                            _enroll(
```

Comment:
Same as above

We don't want to call the underlying `client.link_concession_group_funding_source`, instead this should call `client.update_concession_group_funding_source_expiry`